src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -207,10 +209,78 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+                () => new ProviderAndDocumentToBatchedUpdates());
+
+            private readonly object _gate = new object();
+            private ProviderAndDocumentToBatchedUpdates _idToBatchedUpdates = s_providerPool.Allocate();
+            private Task _updateTask = null;
+
             private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             {
                 this.AssertIsBackground();
-                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
+                if (_disposed)
+                {
+                    return;
+                }
+
+                var document = e.Solution.GetDocument(e.DocumentId);
+                if (document == null)
+                {
+                    // Not a diagnostic event for a document.  Not something we can handle.
+                    return;
+                }
+
+                lock (_gate)
+                {
+                    var key = (e.Id, document);
+                    if (!_idToBatchedUpdates.TryGetValue(key, out var batchedUpdates))
+                    {
+                        batchedUpdates = (removeArgs: null, createArgs: null);
+                    }
+
+                    switch (e.Kind)
+                    {
+                    case DiagnosticsUpdatedKind.DiagnosticsRemoved:
+                        // we're being told about diagnostics going away because a document/project
+                        // was removed.  This supercedes all previous removes and creates for this
+                        // provider+docid.
+                        batchedUpdates.removeArgs = e;
+                        batchedUpdates.createArgs = null;
+                        break;
+                    case DiagnosticsUpdatedKind.DiagnosticsCreated:
+                        // We're creating diagnostics. This supercedes all existing creations for this
+                        // provider +docid.  However, any existing removal for this provider+docid will 
+                        // still happen.
+                        batchedUpdates.createArgs = e;
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(e);
+                    }
+
+                    _idToBatchedUpdates[key] = batchedUpdates;
+
+                    // Check if there's already an in-flight update task.  If so, there's nothing we
+                    // need to do.  The in-flight task will pick up the work we enqueued.  Otherwise,
+                    // create a task to process this work (and anything else that comes in) for 50ms
+                    // from now.
+                    if (_updateTask == null)
+                    {
+                        var token = _owner._listener.BeginAsyncOperation(GetType().Name + "OnDiagnosticsUpdatedOnBackground");
+
+                        _updateTask = Task.Delay(50).ContinueWith(
```

Comment:
what cancel previous update task if same document changed twice?

Reply:
I see so this is not actually batching, it just update every 50ms.

Reply:
> what cancel previous update task if same document changed twice?

No need to cancel.  When the task actually executes, it will handle all the notifications that came in between now and then.

Reply:
yep. I thought you were cancelling previous task and enqueue new one so that it gets delayed again. but this is just putting 50ms buffer on each update.

Reply:
this still don't cancel update on tagger dispose though.

Reply:
That's true.  But worse case that means we just process one more request within 50 ms after tagger dispose. this request will see htat _dispose is set and will immediately bail.  Basically, we have to ask ourselves how important is it to cancel a single request.